### PR TITLE
implement clearcache command

### PR DESCRIPTION
**Pull request by xavierdecoster**
https://github.com/NuGet/Home/issues/1516

cc @emgarten @yishaigalatzer

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        [Option(typeof(NuGetCommand), "ClearCacheCommandClearGlobalPackagesDescription")]
+        public bool ClearGlobalPackages { get; set; }
+
+        public override Task ExecuteCommandAsync()
+        {
+            // Clear the NuGet machine cache
+            if (!string.IsNullOrEmpty(MachineCache.Default?.Source))
+            {
+                Console.WriteLine($"{LocalizedResourceManager.GetString("ClearCacheCommand_ClearingNuGetCache")}: {MachineCache.Default.Source}");
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Most strings have a '{0}' and use LocalizedResourceManager to fill in the parameter instead of concatenating the strings like this is doing. I suggest using that instead to keep things consistent and clear.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        private static void DeleteRecursive(string deletePath)
+        {
+            if (!Directory.Exists(deletePath))
+            {
+                return;
+            }
+
+            foreach (var deleteFilePath in Directory.EnumerateFiles(deletePath).Select(Path.GetFileName))
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Some of these files may be in use, this command should go on and delete as much as possible.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs [hunk 3]
```diff
@@ -256,7 +256,7 @@ public RestoreCommand()
             CheckRequireConsent();
 
             // Run the restore
-            var command = new Commands.RestoreCommand(Console, request);
+            var command = new NuGet.Commands.RestoreCommand(Console, request);
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs:259]: Is this change still needed?

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                // Calling DeleteRecursive rather than Directory.Delete(..., recursive: true)
+                // due to an infrequent exception which can be thrown from that API
+                DeleteRecursive(folderPath);
+
+                Console.WriteLine(LocalizedResourceManager.GetString("ClearCacheCommand_CacheCleared"));
+            }
+            catch (Exception e)
+            {
+                Console.WriteError($"{LocalizedResourceManager.GetString("ClearCacheCommand_UnableToClearCacheDir")}: {e.Message}");
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: I would expect this command to return a non-zero exit code if any exceptions occur during deletes.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx [hunk 5]
```diff
@@ -5245,4 +5245,18 @@ nuget update -Self</value>
   <data name="CommandMSBuildVersion" xml:space="preserve">
     <value>Specifies the version of MSBuild to be used with this command. Supported values are 4, 12, 14. By default the MSBuild in your path is picked, otherwise it defaults to the highest installed version of MSBuild.</value>
   </data>
+  <data name="ClearCacheCommandDescription" xml:space="preserve">
+    <value>Clears your local NuGet cache.</value>
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx]: I would elaborate a bit more here and describe the machine cache and web cache that will be cleared briefly.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx [hunk 6]
```diff
@@ -5245,4 +5245,18 @@ nuget update -Self</value>
...
+  <data name="ClearCacheCommandExamples" xml:space="preserve">
+    <value>nuget clearcache</value>
+    <comment>Please don't localize this string</comment>
+  </data>
+  <data name="ClearCacheCommandSummary" xml:space="preserve">
+    <value>&lt;-clearGlobalPackages&gt;</value>
+    <comment>Please don't localize this string</comment>
+  </data>
+  <data name="ClearCacheCommandClearGlobalPackagesDescription" xml:space="preserve">
+    <value>Also clear the NuGet global packages cache.</value>
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx]: This could be more clear.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                Console.WriteLine($"{LocalizedResourceManager.GetString("ClearCacheCommand_ClearingNuGetHttpCache")}: {httpCacheFolderPath}");
+                ClearCacheDirectory(httpCacheFolderPath);
+            }
+            
+            if (ClearGlobalPackages)
+            {
+                // also clear the global NuGet packages cache
+                var globalPackagesFolderPath = SettingsUtility.GetGlobalPackagesFolder(Settings);
+                Console.WriteLine($"{LocalizedResourceManager.GetString("ClearCacheCommand_ClearingNuGetGlobalPackagesCache")}: {globalPackagesFolderPath}");
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: I like that it lists out the paths :+1:

**Comment by emgarten** [comment]
:watch:

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 1 above)
Comment (deepakaravindr) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Please break long lines

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                Console.WriteLine($"{LocalizedResourceManager.GetString("ClearCacheCommand_ClearingNuGetCache")}: {MachineCache.Default.Source}");
+                ClearCacheDirectory(MachineCache.Default.Source);
+            }
+
+            // Clear NuGet v3 HTTP cache
+            var localAppDataFolderPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            var httpCacheFolderPath = Path.Combine(localAppDataFolderPath, "NuGet", "v3-cache");
+            if (!string.IsNullOrEmpty(httpCacheFolderPath))
+            {
+                Console.WriteLine($"{LocalizedResourceManager.GetString("ClearCacheCommand_ClearingNuGetHttpCache")}: {httpCacheFolderPath}");
```
Comment (deepakaravindr) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Could you use nameof instead of using a string for ClearCacheCommand_ClearingNuGetHttpCache. Same comment for other places

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 2 above)
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: remove the select from the foreach, and just do it inside

Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: We have code for delete recursive already elsewhere https://github.com/NuGet/NuGet.Client/blob/emgarten/321merge/src/NuGet.Core/NuGet.Packaging/PackageExtraction/NuGetPackageUtils.cs#L57

Please merge them into one common utility method

**Comment by yishaigalatzer** [comment]
@bradwilson this one is for you

File: src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs (hunk 3 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs:259]: A clean git checkout on dev was not compiling without that change.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 2 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Talked to @emgarten about this one. Perhaps it's not very convenient to merge the two. 

The commandline needs to act differently when one of the files was unable to be deleted by returning a non-zero response code (see https://github.com/NuGet/NuGet.Client/pull/13#discussion_r41563932). The commandline may want to log these to the console.

Also, more code for recursive deletes here :) https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Dnx.Tooling/Utils/FileOperationUtils.cs#L20

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 7 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Example partial cache clearance: 
image

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 4 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: That means throwing an exception (or adjusting all command exec methods to return a result code). Feel free to review the changes in https://github.com/NuGet/NuGet.Client/commit/ef6950899c19afaa3460a4e3e7faef4b2b54e6f9

File: src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx (hunk 5 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx]: Tried to improve the description in https://github.com/NuGet/NuGet.Client/commit/ef6950899c19afaa3460a4e3e7faef4b2b54e6f9

File: src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx (hunk 6 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx]: Having two system-wide packages folders indeed is confusing :) Tried improving in https://github.com/NuGet/NuGet.Client/commit/ef6950899c19afaa3460a4e3e7faef4b2b54e6f9. Feel free to suggest something better

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 2 above)
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Sounds right, I agree we shouldn't merge in this case.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+            if (commandResult != 0)
+            {
+                throw new CommandLineException(LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_CacheClearFailed)));
+            }
+
+            return Task.FromResult(commandResult);
+        }
+
+        private int ClearCacheDirectory(string folderPath)
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: why not return a bool?

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        {
+            // Calling DeleteRecursive rather than Directory.Delete(..., recursive: true)
+            // due to an infrequent exception which can be thrown from that API
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Please add details to the specific exception being thrown, so the next person maintaining this code get it

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+            {
+                Console.WriteLine(
+                    LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_ClearingNuGetCache)),
+                    MachineCache.Default.Source);
+
+                commandResult = ClearCacheDirectory(MachineCache.Default.Source);
+            }
+
+            // Clear NuGet v3 HTTP cache
+            var localAppDataFolderPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: try this on the kudu console. 

Also this should not compute the path here in the command. This should be a centralized setting/configuration kinda thing just like globalpackagesfolder.

Imagine we allow redirection of that, we want it fixed in one place only.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            else
+            {
+                Console.WriteLine(LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_CacheCleared)));
+                return 0;
+            }
+        }
+
+        private static void DeleteRecursive(string deletePath, List<string> failedDeletes)
+        {
+            if (!Directory.Exists(deletePath))
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Do we protect for stack dive? What happens when this runs on very long paths on *nix?

**Comment by yishaigalatzer** [comment]
:watch: 

Please don't rebase before your next commit, so comments don't disappear.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 9 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Addressed in https://github.com/NuGet/NuGet.Client/commit/2393072e082b4712cdef3904f52be24da66b60ba

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 11 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Moved path logic into SettingsUtility of NuGet.Configuration (where GlobalPackagesFolder code is): https://github.com/NuGet/NuGet.Client/commit/7a50d1ccd7685d43ec1f237fbe38093cdbd8f544

Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Tried on kudu console:
image

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 12 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Are you referring to the case where path length may exceed 260 chars? 
Perhaps this can be mitigated by walking the directory tree upfront, scanning for folders that are too deep, and temporarily moving them to a more shallow location before attempting to delete?

Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: No, I'm referring to potential scenario with hard links that will cause loops. Which is why I don't like writing our own recursive methods for deleting folder.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+                throw new CommandLineException(LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_CacheClearFailed)));
+            }
+
+            return Task.FromResult(0);
+        }
+
+        private bool ClearCacheDirectory(string folderPath)
+        {
+            // Calling DeleteRecursive rather than Directory.Delete(..., recursive: true)
+            // due to an infrequent exception which can be thrown from that API
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: I still don't see a fix for this comment. What infrequent exception is being thrown? And why can't we use the .NET implementation?

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs (hunk 12 above)
Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: I figured somebody must have solved this issue before, so I took a look at how DNU clears the cache.
This piece of code may look familiar :)
https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Dnx.Tooling/Utils/FileOperationUtils.cs#L15

The infrequent exception has to do with write permissions on files in the target (sub)folders. If there are readonly files in the cache location, the following exception will be thrown by `Directory.Delete(..., recursive: true)`: 

>  Exception thrown: 'System.UnauthorizedAccessException' in mscorlib.dll

However, the part that tries to make these files writable seems missing, so I added that: https://github.com/NuGet/NuGet.Client/commit/b87b9e1ce3af03fd0ad24e8321366eb6a41cd96a

Note that I still call into `Directory.Delete(..., recursive: true)` to delete the actual directory: https://github.com/NuGet/NuGet.Client/pull/13/files#diff-16d0285c2ebe50837c5df60fd56b56a1R122

Regarding hard links, `Directory.Delete(..., recursive: true)` will remove the symbolic links but not the actual files that were linked to: https://msdn.microsoft.com/en-us/library/fxeahc5f.aspx

Quoting MSDN:

>  The behavior of this method differs slightly when deleting a directory that contains a reparse point, such as a symbolic link or a mount point. If the reparse point is a directory, such as a mount point, it is unmounted and the mount point is deleted. This method does not recurse through the reparse point. If the reparse point is a symbolic link to a file, the reparse point is deleted and not the target of the symbolic link.
> In some cases, if you have the specified directory open in File Explorer, the Delete method may not be able to delete it.

Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: @yishaigalatzer based on the above and the additional commit, is this change good to go?

Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: No, because you still use recursion into the folder structure. If you want to change the readonly first, you can enumerate all the files recursively and remove the readonly mark first. Then end up with non recursive code, which is a lot cleaner.

Your code will still go into symbolic links as it stands right now See line 120

Comment (xavierdecoster) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs]: Contrary to what's documented on MSDN and what one might expect, `Directory.Delete(folderPath, recursive: true);` does NOT delete any files in the directory tree. Therefore, the recursive method to remove any readonly flags will also call into `File.Delete` for each file.

Now also [explicitly checking](https://github.com/NuGet/NuGet.Client/commit/c9e62da0e341331d9084b6d6ecc8f27a8bc64452#diff-16d0285c2ebe50837c5df60fd56b56a1R148) for the presence of `FileAttributes.ReparsePoint` on directories.

Please review commits https://github.com/NuGet/NuGet.Client/commit/c9e62da0e341331d9084b6d6ecc8f27a8bc64452 and https://github.com/NuGet/NuGet.Client/commit/ccd4ea684fb978a2f8deeb8ef38614e460dbe11f

**Comment by xavierdecoster** [comment]
Will close this PR and send a new one to keep track of this discussion.
